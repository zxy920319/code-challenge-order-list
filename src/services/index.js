import axios from "axios";

const ENDPOINT = "https://api.myjson.com/bins";

const handleError = err => {
  // eslint-disable-next-line no-console
  console.warn(err);
  return false;
};

export async function GetUserInfo() {
  try {
    const res = await axios.get(ENDPOINT + "/pdefl");
    return res.data;
  } catch (err) {
    return handleError(err);
  }
}

export async function GetLastOrders() {
  try {
    const res = await axios.get(ENDPOINT + "/19f9bd");
    return res.data;
  } catch (err) {
    return handleError(err);
  }
}

export async function GetOrderInfo() {
  try {
    const res = await axios.get(ENDPOINT + "/o1sp5");
    return res.data;
  } catch (err) {
    return handleError(err);
  }
}
