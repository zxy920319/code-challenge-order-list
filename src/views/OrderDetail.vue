<template>
  <v-scale-transition>
    <v-row
      v-if="order"
      justify="center"
      align="center"
      class="order-detail-constainer"
    >
      <!-- <v-col cols="10" sm="8">
        <v-row justify="end" class="ma-0">
          <v-btn depressed color="accent">
            Back
          </v-btn>
        </v-row>
			</v-col>-->
      <v-col cols="10" sm="8">
        <v-card dark color="tertiary">
          <v-card-title class="headline">{{
            order.status.currentStatus
          }}</v-card-title>

          <v-card-text class="order-info">
            <div>
              <div class="subtitle-2">Order Time</div>
              <p class="body-2">{{ formatTime(order.orderTime) }}</p>
            </div>
            <div>
              <div class="subtitle-2">Restaurant Name</div>
              <p class="body-2">{{ order.restaurantName }}</p>
            </div>
            <div>
              <div class="subtitle-2">Order Total</div>
              <p class="body-2">{{ order.orderTotal + currency }}</p>
            </div>
            <div>
              <div class="subtitle-2">Delivery Costs</div>
              <p class="body-2">{{ order.deliveryCosts + currency }}</p>
            </div>
            <div>
              <div class="subtitle-2">Payment Method</div>
              <p class="body-1">{{ order.paymentMethod }}</p>
            </div>
          </v-card-text>
        </v-card>
      </v-col>
      <v-col cols="10" sm="8">
        <v-row class="ma-0 flex-wrap">
          <v-card
            v-for="(item, index) in order.products"
            :key="index"
            class="food-info"
            outlined
          >
            <v-list-item three-line>
              <v-list-item-content>
                <div class="overline mb-4">NUMBER: {{ item.amount }}</div>
                <v-list-item-title class="title mb-1">{{
                  item.productName
                }}</v-list-item-title>
                <v-list-item-subtitle>{{
                  item.productDescription
                }}</v-list-item-subtitle>
              </v-list-item-content>

              <v-list-item-avatar tile size="80">
                <img :src="require('@/assets/img/food.png')" alt="John" />
              </v-list-item-avatar>
            </v-list-item>
          </v-card>
        </v-row>
      </v-col>
      <v-col v-if="$vuetify.breakpoint.mdAndUp" cols="10" sm="8" class="mb-8">
        <v-timeline>
          <v-timeline-item
            v-for="(item, index) in order.status.history"
            :key="index"
            color="primary lighten-2"
            fill-dot
            :right="Boolean(index % 2)"
            :left="!Boolean(index % 2)"
          >
            <v-card>
              <v-card-title class="primary lighten-2">
                <v-icon dark size="42" class="mr-4">mdi-magnify</v-icon>
                <h2 class="headline white--text font-weight-light">
                  {{ item.status }}
                </h2>
              </v-card-title>
              <v-container>
                <v-row>
                  <v-col cols="12" md="10">{{
                    formatTime(item.statusStart)
                  }}</v-col>
                </v-row>
              </v-container>
            </v-card>
          </v-timeline-item>
        </v-timeline>
      </v-col>
      <v-col v-else cols="10" class="mb-8">
        <v-timeline align-top dense>
          <v-timeline-item
            v-for="(item, index) in order.status.history"
            :key="index"
            color="primary lighten-2"
            small
          >
            <v-card>
              <v-card-text>
                <v-row class="pt-1">
                  <v-col>
                    <strong>{{ item.status }}</strong>
                    <div class="caption">
                      {{ formatTime(item.statusStart) }}
                    </div>
                  </v-col>
                </v-row>
              </v-card-text>
            </v-card>
          </v-timeline-item>
        </v-timeline>
      </v-col>
    </v-row>
  </v-scale-transition>
</template>

<script>
import { GetOrderInfo } from "@/services";
import * as moment from "moment";
import mixin from "@/mixins";

export default {
  name: "OrderDetail",

  mixins: [mixin],

  data: () => ({
    order: null
  }),

  mounted() {
    this.getOrderInfo();
  },

  methods: {
    async getOrderInfo() {
      this.order = await GetOrderInfo();
    },

    formatTime(time) {
      if (!time) return "";
      return moment(time).format("h:mm:ss a, Do MMMM YYYY");
    }
  }
};
</script>
