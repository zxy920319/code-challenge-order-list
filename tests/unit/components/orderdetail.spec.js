import Vuetify from "vuetify";
import Vuex from "vuex";
import { shallowMount, createLocalVue } from "@vue/test-utils";
import OrderDetail from "@/views/OrderDetail.vue";
import { GetOrderInfo } from "@/services";

jest.mock("@/services");

const mockData = {
  orderReference: "FVVAWO",
  customerId: 8481133476,
  orderTime: "2019-07-15 16:44:06",
  restaurantName: "Big Burger Joint",
  orderTotal: "16.95",
  deliveryCosts: "2.75",
  paymentMethod: "iDeal",
  deliveryAddress: "Streetname 1",
  status: {
    currentStatus: "In transit",
    history: [
      {
        status: "Food is being prepared",
        statusStart: "2019-07-15 16:49:45",
        statusEnd: "2019-07-15 16:59:54"
      },
      {
        status: "Order is accepted",
        statusStart: "2019-07-15 16:45:23",
        statusEnd: "2019-07-15 16:49:45"
      },
      {
        status: "Order is placed",
        statusStart: "2019-07-15 16:44:06",
        statusEnd: "2019-07-15 16:45:23"
      }
    ]
  },
  products: [
    {
      amount: 1,
      productName: "Cheeseburger with fries on the side",
      productDescription:
        "250gr beef burger with cheddar cheese. Topped with lettuce, tomato and unions.",
      pricePerUnitWithTax: "9.45"
    },
    {
      amount: 1,
      productName: "Big Mega Hotdog",
      productDescription:
        "(Mustard and ketchup) With bacon, pickels, union, cheddar and grilled hotdog.",
      pricePerUnitWithTax: "4.75"
    }
  ]
};
GetOrderInfo.mockResolvedValue(mockData);

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(Vuetify);

let mutation = {
  setProfile: state => (state.profile = {})
};

let store = new Vuex.Store({
  state: { profile: {} },
  mutation
});

let vuetify = new Vuetify();

jest.mock("moment", () => () => ({
  format: () => "2018–01–30T12:34:56+00:00"
}));

describe("OrderDetail", () => {
  const wrapper = shallowMount(OrderDetail, {
    store,
    vuetify,
    localVue,
    stubs: {
      "v-scale-transition": false
    }
  });

  test("GetOrderInfo will be called", () => {
    expect(GetOrderInfo.mock.calls.length).toBe(1);
  });

  it("should render", () => {
    wrapper.setData({ order: mockData });
    expect(wrapper.vm.order).toBeTruthy();
  });
});
