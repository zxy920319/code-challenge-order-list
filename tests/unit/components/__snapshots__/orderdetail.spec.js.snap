// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderDetail should render class order-detail-constainer 1`] = `
<v-scale-transition-stub origin="top center 0">
  <!---->
</v-scale-transition-stub>
`;
